rule/unused-param.go

```diff
@@ -0,0 +1,97 @@
+package rule
+
+import (
+	"fmt"
+	"go/ast"
+
+	"github.com/mgechev/revive/lint"
+)
+
+// UnusedParamRule lints unused params in functions.
+type UnusedParamRule struct{}
+
+// Apply applies the rule to given file.
+func (r *UnusedParamRule) Apply(file *lint.File, _ lint.Arguments) []lint.Failure {
+	var failures []lint.Failure
+
+	onFailure := func(failure lint.Failure) {
+		failures = append(failures, failure)
+	}
+
+	w := lintUnusedParamRule{onFailure: onFailure}
+
+	ast.Walk(w, file.AST)
+
+	return failures
+}
+
+// Name returns the rule name.
+func (r *UnusedParamRule) Name() string {
+	return "unused-parameter"
+}
+
+type lintUnusedParamRule struct {
+	onFailure func(lint.Failure)
+}
+
+func (w lintUnusedParamRule) Visit(node ast.Node) ast.Visitor {
+	switch n := node.(type) {
+	case *ast.FuncDecl:
+		fv := funcVisitor{params: retrieveNamedParams(n.Type.Params.List)}
+		if n.Body != nil {
+			ast.Walk(fv, n.Body)
+			checkUnusedParams(w, fv.params, n)
+		}
+		return nil
+	}
+
+	return w
+}
+
+type funcVisitor struct {
+	params map[string]bool
+}
+
+func (v funcVisitor) Visit(node ast.Node) ast.Visitor {
+	switch n := node.(type) {
+	case *ast.Ident:
+		if n.Obj != nil {
+			if n.Obj.Kind.String() == "var" {
+				markParamAsUsed(n, v.params)
+			}
+		}
+	}
+
+	return v
+}
+
+func retrieveNamedParams(pl []*ast.Field) map[string]bool {
+	result := make(map[string]bool, len(pl))
+	for _, p := range pl {
+		for _, n := range p.Names {
+			if n.Name != "_" {
+				result[n.Name] = true
+			}
+		}
+	}
+	return result
+}
+
+func checkUnusedParams(w lintUnusedParamRule, params map[string]bool, n *ast.FuncDecl) {
+	for k, v := range params {
+		if v {
+			w.onFailure(lint.Failure{
+				Confidence: 0.8, // confidence is not 1.0 because of shadow variables
```

Comment:
How hard do you think it'll be to take into consideration variable shadowing?
- We already have the base case (no local variables with same name & parameter not used)
- It's easy if we find that given parameter is used before there's a declaration of a variable with the same name
- While visiting blocks, we should keep track of the lexical scope as a stack of sets (or `map[string]bool`):
    - If there's no local variable with the same name as a parameter used in any nested block, we don't mark it (we traverse the stack and check if a variable in any block with the same name as the param exists)
    - Otherwise, we do

A complication I see are short forms, such as:

```go
if param := fn(); predicate(param) {
  // do stuff
}
```

Am I missing something?

Reply:
Following your suggestions, I'm working in a version that tracks scopes. The handling of short forms seems to be not so hard.

Reply:
Please check the last committed version. It handles cases like
```go
func f0(param int) {
	param := param
}

func f1(param int) { // MATCH /parameter 'param' seems to be unused, consider removing or renaming it as _/
	if param := fn(); predicate(param) {
		// do stuff
	}
}

func f2(param int) { // MATCH /parameter 'param' seems to be unused, consider removing or renaming it as _/
	switch param := fn(); param {
	default:
		
	}
}

func f3(param myStruct) { 
	a:= param.field
}

func f4(param myStruct, c int) { // MATCH /parameter 'c' seems to be unused, consider removing or renaming it as _/ 
	param.field = "aString"
	param.c = "sss" 
}
```